### PR TITLE
T/858 Refactoring: improve batch API

### PR DESCRIPTION
**Pull request by pjasiun**
Refactoring: improve batch API. Make it possible to use batch API on detached nodes. Use batch API instead of elements API in view to model conversion.

### CODE REVIEW

File: src/conversion/viewconversiondispatcher.js [hunk 1]
```diff
@@ -203,24 +208,23 @@ export default class ViewConversionDispatcher {
...
+
+		for ( const viewChild of Array.from( input.getChildren() ) ) {
+			const modelChild = this._convertItem( viewChild, consumable, additionalData );
+
+			if ( modelChild instanceof ModelNode ) {
+				batch.append( modelChild, documentFragment );
+			} else if ( modelChild instanceof ModelDocumentFragment ) {
+				for ( const child of Array.from( modelChild ) ) {
+					batch.append( child, documentFragment );
+				}
```
Comment (pjasiun) [inline src/conversion/viewconversiondispatcher.js]: Can't we just appen document fragment to another document fragment?

File: src/model/documentselection.js [hunk 2]
```diff
@@ -723,15 +723,15 @@ function clearAttributesStoredInElement( changes, batch, document ) {
 
 	// `changes.range` is not set in case of rename, root and marker operations.
 	// None of them may lead to the element becoming non-empty.
-	if ( !changeParent || changeParent.isEmpty ) {
+	if ( !changeParent || changeParent.is( 'documentFragment' ) || changeParent.isEmpty ) {
```
Comment (pjasiun) [inline src/model/documentselection.js]: Is it still needed? I see now change is fired only for docment changes.

Comment (oskarwrobel) [reply src/model/documentselection.js]: I think you are right.

**Comment by pjasiun** [comment]
In: https://github.com/ckeditor/ckeditor5-typing/compare/master...t/ckeditor5-engine/858

```
this._buffer.batch.insert( paragraph, limitElement );
```
Is not correct API usage. Should be:
```
this._buffer.batch.insert( paragraph, limitElement, 0 );
```
Or:
```
this._buffer.batch.append( paragraph, limitElement );
```
(does not matter if you insert at the beginning or at the end, the element is empty)

**Comment by pjasiun** [comment]
Same in `ckeditor5-image` (`conversionApi.batch.insert( modelChildren, modelImage );`).

**Comment by pjasiun** [comment]
https://github.com/ckeditor/ckeditor5-image/compare/master...t/ckeditor5-engine/858#diff-f91836fb3901004a578ed9233b54df72R177

Shouldn't be:
```
batch.appendElement( 'caption', item, 'end' );
```
(attributes parameter is optional, but the position was at the end)

**Comment by oskarwrobel** [comment]
This url doesn't work but append is always at the end.

**Comment by pjasiun** [comment]
Nevermid, my bad.

**Pull request merged**